pkg/providers/docker/docker.go

```diff
@@ -463,10 +463,26 @@ func (p *provider) RunPostControlPlaneCreation(ctx context.Context, clusterSpec
 	return nil
 }
 
-func buildMapForWorkerNodeGroupsByName(prevWorkerNodeGroups []v1alpha1.WorkerNodeGroupConfiguration) map[string]v1alpha1.WorkerNodeGroupConfiguration {
-	prevConfigs := make(map[string]v1alpha1.WorkerNodeGroupConfiguration, len(prevWorkerNodeGroups))
-	for _, config := range prevWorkerNodeGroups {
-		prevConfigs[config.Name] = config
+func buildMapForWorkerNodeGroupsByName(workerNodeGroups []v1alpha1.WorkerNodeGroupConfiguration) map[string]v1alpha1.WorkerNodeGroupConfiguration {
+	workerNodeGroupConfigs := make(map[string]v1alpha1.WorkerNodeGroupConfiguration, len(workerNodeGroups))
+	for _, config := range workerNodeGroups {
+		workerNodeGroupConfigs[config.Name] = config
 	}
-	return prevConfigs
+	return workerNodeGroupConfigs
+}
+
+func (p *provider) NodeGroupsToDelete(ctx context.Context, workloadCluster *types.Cluster, currentSpec, newSpec *cluster.Spec) ([]*clusterv1.MachineDeployment, error) {
+	workerConfigs := buildMapForWorkerNodeGroupsByName(newSpec.Spec.WorkerNodeGroupConfigurations)
+	machineDeployments := make([]*clusterv1.MachineDeployment, 0, len(currentSpec.Spec.WorkerNodeGroupConfigurations))
+	for _, prevWorkerNodeGroupConfig := range currentSpec.Spec.WorkerNodeGroupConfigurations {
+		if _, ok := workerConfigs[prevWorkerNodeGroupConfig.Name]; !ok {
+			workerNodeGroupName := fmt.Sprintf("%s-%s", workloadCluster.Name, prevWorkerNodeGroupConfig.Name)
+			machineDeployment, err := p.providerKubectlClient.GetMachineDeployment(ctx, workerNodeGroupName, executables.WithKubeconfig(workloadCluster.KubeconfigFile), executables.WithNamespace(constants.EksaSystemNamespace))
+			if err != nil {
+				return nil, err
+			}
+			machineDeployments = append(machineDeployments, machineDeployment)
+		}
+	}
+	return machineDeployments, nil
```

Comment:
There is a lot of duplicated code in between providers for this method. Since the only difference (if any) should be the naming, what do you think about:
* Move the logic to determine the removed worker node groups somewhere else. Maybe in the `cluster` package. Just a standalone method that takes a cluster config and returns a slice of worker node group configs.
* Return only the machine deployment names from this method and move the `GetMachineDeployment` logic to cluster manager.